hubblestack_nova/modules/openssl.py

```diff
@@ -0,0 +1,283 @@
+    port:
+        - the port is required only if both:
+            - the endpoint is defined
+            - https is configured on another port the 443 on the endpoint
+        - WARNING: if the port is not the on configured for https on the endpoint, downloading the certificate from
+          the endpoint will timeout and the check will be failed
+        - if endpoint is defined but the port is not, the module will try, by default, to use port 443
+    notAfter:
+        - the minimum number of days left until the certificate should expire
+        - if the certificate will expire in less then the value given here, the check will fail
+        - if notAfter is missing, the default value is 0; basically, the if the expiration date is in the future, the
+          check will be passed
+    notBefore:
+        - the expected number of days until the certificate becomes valid
+        - this is useful only if you expect the certificate to be valid after a certain date
+        - if missing, 0 is the default value (read more bellow)
+    fail_if_notBefore:
+        - if True, the check will fail only if notBefore is 0 (or missing): if the certificate is not valid yet, but
+          it is expected to be
+        - the default value is False - the check will fail only if the certificate expiration date is valid
+
+Some notes:
+    - if BOTH file and endpoint are present / missing, the check will fail; only one certificate has to be present for
+      each check
+    - the YAML supports also the control key, just as the other modules do
+
+Known issues: for unknown reasons (yet), the module can fail downloading the certificate from certain endpoints. When
+this happens, the check will be failed.
+
+'''
+
+
+from __future__ import absolute_import
+import logging
+
+import fnmatch
+import copy
+import salt.utils
+import datetime
+import time
+
+import ssl
+
+try:
+    import OpenSSL
+    _HAS_OPENSSL = True
+except ImportError:
+    _HAS_OPENSSL = False
+
+log = logging.getLogger(__name__)
+
+__tags__ = None
+__data__ = None
+
+def __virtual__():
+    if salt.utils.is_windows():
+        return False, 'This audit module only runs on linux'
+    if not _HAS_OPENSSL:
+        return (False, 'The python-OpenSSL library is missing')
+    return True
+
+
+def audit(data_list, tags, verbose=False):
+    __data__ = {}
+    for data in data_list:
+        _merge_yaml(__data__, data)
+    __tags__ = _get_tags(__data__)
+
+    log.trace('service audit __data__:')
+    log.trace(__data__)
+    log.trace('service audit __tags__:')
+    log.trace(__tags__)
+
+    ret = {'Success': [], 'Failure': [], 'Controlled': []}
+    for tag in __tags__:
+        if fnmatch.fnmatch(tag, tags):
+            for tag_data in __tags__[tag]:
+                if 'control' in tag_data:
+                    ret['Controlled'].append(tag_data)
+                    continue
+
+                endpoint = tag_data.get('endpoint', None)
+                file = tag_data.get('file', None)
```

Comment:
Can we rename this variable? In python 2, `file` is a builtin function that we shouldn't mask this way.